proxy/proxy/client_test.go

```diff
@@ -183,3 +192,43 @@ func TestMaximumConnectionsCount(t *testing.T) {
 		t.Errorf("client should have dialed exactly the maximum of %d connections (%d connections, %d dials)", maxConnections, numConnections, dials)
 	}
 }
+
+func TestRefreshTimer(t *testing.T) {
+	refreshCertBuffer = time.Second
+	timeToExpire := 2 * time.Second
+	b := &fakeCerts{}
+	c := &Client{
+		Certs: &blockingCertSource{
+			map[string]*fakeCerts{
+				instance: b,
+			},
+			time.Now().Add(timeToExpire),
+		},
+		Dialer: func(string, string) (net.Conn, error) {
+			return nil, errFakeDial
+		},
+		RefreshCfgThrottle: 500 * time.Millisecond,
+	}
+
+	// Call Dial to cache the cert.
+	if _, err := c.Dial(instance); err != errFakeDial {
+		t.Errorf("unexpected error: %v", err)
+	}
+
+	cached, ok := c.cfgCache[instance]
```

Comment:
We should strive to write all of our code to be free of race conditions, especially tests: when we run them we can enable the race detector to check for problems, and ideally any problems found would be in the SUT rather than the unit tests themselves!

I suggest running the tests under the race detector (add `-race` to `go test`) until it doesn't report problems.

(and if there happen to be problems already, pleas file an issue! :] we don't run these tests regularly under the race detector so it's possible another regression has slipped in)

Reply:
Done.